pages/partials/navbar.ejs

```diff
@@ -9,140 +20,420 @@
-            <a class="dropdown-item" href="<%= plainUrlPrefix %>/settings">Settings</a>
-            <% if (typeof authz_data !== 'undefined' && authz_data.authn_has_instructor_view) { %>
-            <a class="dropdown-item" href="<%= urlPrefix %>/effectiveUser">Change effective user</a>
-            <% } %>
-            <a class="dropdown-item news-item-link" href="<%= urlPrefix %>/news_items">
-              News
-              <% if (typeof news_item_notification_count != 'undefined' && news_item_notification_count > 0) { %>
-                <span class="badge badge-pill badge-primary news-item-link-count"><%= news_item_notification_count %></span>
+        <div class="dropdown-menu dropdown-menu-right" aria-labelledby="navbarDropdown">
+
+          <% if (locals.authn_is_administrator) { %>
+
+            <button class="dropdown-item" type="button" name="administrator">
+              <% if (locals.access_as_administrator) { %>Turn off administrator access<% } else { %>Turn on administrator access<% } %>
+            </button>
+
+            <div class="dropdown-divider"></div>
+
+          <% } %>
+
+          <% if (locals.course_instance && (authz_data.authn_has_course_permission_preview || authz_data.authn_has_course_instance_permission_view)) {
+            // Only show "View type" menu(s) if the following two things are true:
+            // - The authn user was given access a course instance (so, both viewType and authz_data also exist).
+            // - In particular, the authn user has instructor access to this course instance.
+            //
+            // Note that the effective user may still have been denied access. In this
+            // case, urlPrefix may not be consistent with the page that the effective user
+            // was trying to access (instead, it will be consistent with a "plain" page).
+            // So, to be safe, we use plainUrlPrefix and construct all full URLs by hand.
+
+            var instructorLink = '#';
+            var studentLink = '#';
+            if (viewType == 'instructor') {
+              if (locals.assessment && assessment.id) {
+                studentLink = `${plainUrlPrefix}/course_instance/${course_instance.id}/assessment/${assessment.id}`;
+              } else {
+                studentLink = `${plainUrlPrefix}/course_instance/${course_instance.id}/assessments`;
+              }
+            } else {
+              if (locals.question && question.id) {
+                instructorLink = `${urlPrefix}/instructor/question/${question.id}`;
+              } else if (locals.assessment_instance && assessment_instance.assessment_id) {
+                instructorLink = `${urlPrefix}/instructor/assessment/${assessment_instance.assessment_id}`;
+              } else {
+                instructorLink = `${urlPrefix}/instructor/instance_admin`;
+              }
+            }
+
+            var headingAuthnViewTypeMenu = 'View type';
```

Comment:
Maybe. Things get a little tricky with effective users - see below. It's not clear that a "View as" construction would be better, with effective users, than the current "View type". Let me know if you have a suggestion - this could also be an easy UI improvement for a future PR.

![Screen Shot 2021-01-12 at 11 27 58 AM](https://user-images.githubusercontent.com/14169175/104349962-3dfbbd80-54c9-11eb-8952-2636e63f3dfc.png)

Note also that keeping "View type" and "... page ..." tries to make clear that this involves changing the _page_ in addition to changing the way you view the page. The "View as" construction, without "... page ...", may make this less clear.

Reply:
Ah, I hadn't realized what this looks like with an effective user. In the screenshot above, why do we need to have two different places where the effective user is shown (in the yellow box and in the "view type" list)? Is it the case that choosing one of the "view type for dev user" options is the same as clicking "remove" on the yellow box?

I like the yellow box a lot. I'm wondering whether we could have just one "view type" (or "view as") section that wasn't specific to the user, and the yellow box to control the user? I guess that at the moment it feels redundant to me rather than separating "view type" (three menu items) from effective user (the yellow box).

Reply:
First, note that "View type" can have different options for the real user and the effective user, for example when the real user is an instructor and the effective user is not:

![Screen Shot 2021-01-12 at 1 59 56 PM](https://user-images.githubusercontent.com/14169175/104366219-a5703800-54de-11eb-8238-d51bd4c0a96f.png)

Second, note that "View type" can change _both_ the page and the effective user. Clicking on "Instructor page" for the Dev User in this example will remove the `UID`, `Course role`, and `Course instance role` overrides, but it will also change the page. Furthermore, it will _not_ remove other overrides - if we were also in exam mode for this example, we would remain in exam mode when we click "Instructor page."

I am ok with erring on the side of redundancy in this case - so, exposing "Instructor page" for _both_ `Dev User` and `Student One` in this example - although there may be a much better design than the one I have currently that does not feel similarly redundant.

I am sympathetic with wanting to strictly separate "view" from "user" and tried many different ways of doing so - I don't think it's easy, which is why I arrived at what I did.

I may be misunderstanding some of what you suggest because it is very easy to mix up words like "user", "Effective user", and "UID". For example, when you say "[not] specific to the user," perhaps you actually mean "UID" and not "user."

Reply:
I somehow ended up commenting in two different threads here about essentially the same issue. I'll try and consolidate here. Copying from below (edited for clarity):

> The main UI issue I'm worried about it is making it clear to users which view they are in, and where to go to change the view. In the current `master` we have these huge obnoxious Instructor/Student buttons in the navbar. I like getting rid of those, but I fear we might have gone too far in hiding everything under the user's name with no other indications of what's going on. Maybe we could just change the navbar item to be "User Name (instructor)" where the "(instructor)" text changes to reflect the view type?

This conversation helped me understand my real concerns. I think I'm currently worried about:

1. For existing users who are used to the "Instructor"/"Student" buttons in the navbar, how can we help them realize that this functionality moved into the "User Name" menu?
2. For new users, how do we indicate that "View this page as a student" functionality is in the "User Name" menu? I worry that this is hard to discover and many people won't realize that they can view as a student.
3. If an instructor is in a student view, there is currently no indication on the page (without clicking the "User Name" menu) about what's going on and how to get back to their regular view.
4. I feel like "Student page" sounds like it's referring to a specific page, rather than to the student version/view of the current page. Perhaps change to "Student view"/"Instructor view"?
5. I worry that it might not be clear what "without/with access rules" mean. Maybe "Student view - unrestricted access" and "Student view - restricted by access rules"?

Reply:
> I am sympathetic with wanting to strictly separate "view" from "user" and tried many different ways of doing so - I don't think it's easy, which is why I arrived at what I did.

This makes sense to me and I believe you that it's a hard problem 😄. I'm happy to defer to your solution here.

Reply:
I partially addressed this in https://github.com/PrairieLearn/PrairieLearn/pull/3020#issuecomment-896145521 and I added a Todo at the top of the PR to add a badge that draws the user's attention to the user-name menu in the top left of the navbar.